scalding-core/src/main/scala/com/twitter/scalding/Mode.scala

```diff
@@ -63,72 +75,258 @@ object Mode {
+
+  /**
+   * Verify whether sources are valid, given the applicable source strictness mode
+   *
+   * @throws InvalidSourceException if sources are missing
+   */
+  def validateTap(schemedSource: SchemedSource): Unit
+
+  /**
+   * Create a memory-only tap for the purpose of spooling some tuples from memory.
+   * On Local modes, it is also possible to write to a memory tap.
+   */
+  def createMemoryTap(readOrWrite: AccessMode, fields: Fields, asBuffer: mutable.Buffer[Tuple]): Tap[_, _, _]
+
+  /** Create a write-only tap which simply discards the results */
+  def createNullTap: Tap[_, _, _]
+
+  // access this through the DistributedCacheFile object
+  private[scalding] def addCachedFile(file: UncachedFile): CachedFile
+
+  /** Read the content of a file into a string, UTF-8 encoding is assumed */
+  def readFromFile(filename: String): String
+  /** Read the content of a file into a string, UTF-8 encoding is assumed */
+  def writeToFile(filename: String, text: String): Unit
+}
+
+object StorageMode {
+  trait TemporarySource[T] {
+    def sink(conf: Config): TypedSink[T]
+    def downstreamPipe(conf: Config): TypedPipe[T]
+  }
+}
+
+trait Mode extends java.io.Serializable {
+  def storageMode: StorageMode
+  def executionMode: ExecutionMode
+
+  def strictSources: Boolean
+  def name: String
+
+  // Override this if you define new testing modes
+  def isTesting: Boolean = false
+
+  /**
+   * Using a new FlowProcess, which is only suitable for reading outside
+   * of a map/reduce job, open a given tap and return the TupleEntryIterator
+   */
+  final def openForRead(config: Config, tap: Tap[_, _, _]): TupleEntryIterator = executionMode.openForRead(config, tap)
+
+  /** Create a new FlowConnector for this cascading planner */
+  protected // will probably have to end up just "@deprecated"
+  final def newFlowConnector(config: Config): FlowConnector = newFlowConnector(config)
+
+  /** Create a new Flow for this cascading planner */
+  def newFlow(config: Config, flowDef: FlowDef): Flow[_] = executionMode.newFlow(config, flowDef)
+
+  /**
+   * @return true if the file exists on the current filesystem.
+   */
+  final def fileExists(filename: String): Boolean = storageMode.fileExists(filename)
+
   @deprecated("A Config is needed, especially if any kryo serialization has been used", "0.12.0")
   final def openForRead(tap: Tap[_, _, _]): TupleEntryIterator =
     openForRead(Config.defaultFrom(this), tap)
+}
 
-  // Returns true if the file exists on the current filesystem.
-  def fileExists(filename: String): Boolean
-  /** Create a new FlowConnector for this cascading planner */
-  def newFlowConnector(props: Config): FlowConnector
+/**
+ * Any Mode that runs using local resources only
+ */
+trait LocalMode extends Mode {
+}
+
+/**
+ * Any Mode that runs over a cluster (as opposed to Local) is a ClusterMode.
+ */
+trait ClusterMode extends Mode {
 }
 
-trait HadoopMode extends Mode {
```

Comment:
we have user code that assumes this. Can we find a way not to break those folks?